pydantic_ai_slim/pydantic_ai/durable_exec/temporal/_dynamic_toolset.py

```diff
@@ -0,0 +1,155 @@
+                    )
+
+                args_dict = tool.args_validator.validate_python(params.tool_args)
+                return await self._wrap_call_tool_result(self.wrapped.call_tool(params.name, args_dict, ctx, tool))
+
+        call_tool_activity.__annotations__['deps'] = deps_type
+
+        self.call_tool_activity = activity.defn(name=f'{activity_name_prefix}__dynamic_toolset__{self.id}__call_tool')(
+            call_tool_activity
+        )
+
+    @property
+    def temporal_activities(self) -> list[Callable[..., Any]]:
+        return [self.get_tools_activity, self.call_tool_activity]
+
+    async def __aenter__(self) -> Self:
+        return self
+
+    async def __aexit__(self, *args: Any) -> bool | None:
+        return None
+
+    async def get_tools(self, ctx: RunContext[AgentDepsT]) -> dict[str, ToolsetTool[AgentDepsT]]:
+        if not workflow.in_workflow():
+            return await super().get_tools(ctx)
+
+        serialized_run_context = self.run_context_type.serialize_run_context(ctx)
+        tool_defs = await workflow.execute_activity(
+            activity=self.get_tools_activity,
+            args=[
+                _GetToolsParams(serialized_run_context=serialized_run_context),
+                ctx.deps,
+            ],
+            **self.activity_config,
+        )
+        return {name: self._tool_for_tool_def(tool_def) for name, tool_def in tool_defs.items()}
+
+    async def call_tool(
+        self,
+        name: str,
+        tool_args: dict[str, Any],
+        ctx: RunContext[AgentDepsT],
+        tool: ToolsetTool[AgentDepsT],
+    ) -> Any:
+        if not workflow.in_workflow():
+            return await super().call_tool(name, tool_args, ctx, tool)
+
+        tool_activity_config = self.tool_activity_config.get(name)
+        if tool_activity_config is False:
+            return await super().call_tool(name, tool_args, ctx, tool)
+
+        merged_config = self.activity_config | (tool_activity_config or {})
+        serialized_run_context = self.run_context_type.serialize_run_context(ctx)
+        return self._unwrap_call_tool_result(
+            await workflow.execute_activity(
+                activity=self.call_tool_activity,
+                args=[
+                    CallToolParams(
+                        name=name,
+                        tool_args=tool_args,
+                        serialized_run_context=serialized_run_context,
+                        tool_def=tool.tool_def,
+                    ),
+                    ctx.deps,
+                ],
+                **merged_config,
+            )
+        )
+
+    def _tool_for_tool_def(self, tool_def: ToolDefinition) -> ToolsetTool[AgentDepsT]:
+        """Create a ToolsetTool from a ToolDefinition for use outside activities."""
+        return ToolsetTool(
+            toolset=self,
+            tool_def=tool_def,
+            max_retries=1,
+            args_validator=TOOL_SCHEMA_VALIDATOR,
```

Comment:
Let's add a note to say that this is fine, because we run the actual args validator inside `call_tool_activity`